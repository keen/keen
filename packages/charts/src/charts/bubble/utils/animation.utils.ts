export const randomizeVector = () => {
  const min = -10;
  const max = 10;
  return [
    Math.floor(Math.random() * (max - min + 1)) + min,
    Math.floor(Math.random() * (max - min + 1)) + min,
  ];
};
