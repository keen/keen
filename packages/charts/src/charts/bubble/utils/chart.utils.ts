import { scaleLinear, scaleOrdinal } from 'd3-scale';
import {
  calculateRange,
  calculateScaleDomain,
  getPaletteColor,
} from '@keen.io/charts-utils';

import { Dimension, Margins } from '../../../types';

import { Bubble } from '../types';

type Options = {
  data: Record<string, any>[];
  dimension: Dimension;
  margins: Margins;
  colors: string[];
  minAreaRadius: number;
  maxAreaRadius: number;
  valueKey: string;
  xDomainKey: string;
  yDomainKey: string;
  labelSelector: string;
  disabledKeys: string[];
};

const createScale = ({
  data,
  range,
  key,
}: {
  data: Record<string, any>[];
  range: [number, number];
  key: string;
}) => {
  const { minimum, maximum } = calculateRange(data, 'auto', 'auto', [key]);
  const scale = scaleLinear().domain([minimum, maximum]).range(range);
  calculateScaleDomain(scale, minimum, maximum);

  return {
    scale,
    middlePoint: scale(maximum / 2),
  };
};

export const generateBubbles = ({
  data,
  margins,
  dimension,
  colors,
  labelSelector,
  disabledKeys,
  valueKey,
  xDomainKey,
  yDomainKey,
  minAreaRadius,
  maxAreaRadius,
}: Options) => {
  const bubbles: Bubble[] = [];

  const { scale: xScale, middlePoint: xScalePoint } = createScale({
    data,
    key: xDomainKey,
    range: [margins.left, dimension.width - margins.right],
  });

  const { scale: yScale, middlePoint: yScalePoint } = createScale({
    data,
    key: yDomainKey,
    range: [dimension.height - margins.bottom, margins.top],
  });

  const { minimum, maximum } = calculateRange(data, 'auto', 'auto', [valueKey]);
  const radiusScale = scaleLinear()
    .domain([minimum, maximum])
    .range([minAreaRadius, maxAreaRadius]);

  data.forEach((item, idx) => {
    const label = item[labelSelector];
    const color = getPaletteColor(idx, colors);
    if (disabledKeys && !disabledKeys.includes(label)) {
      const bubble = {
        key: `${label}-${idx}`,
        x: xScale(item[xDomainKey]),
        y: yScale(item[yDomainKey]),
        color,
        selector: [idx],
        radius: radiusScale(item[valueKey]),
      };
      bubbles.push(bubble);
    }
  });

  return {
    bubbles: bubbles.sort((a, b) => b.radius - a.radius),
    xScale,
    yScale,
    middlePoint: [xScalePoint, yScalePoint] as [number, number],
  };
};

export const bubbleColorScale = (colors: string[]) => scaleOrdinal(colors);
