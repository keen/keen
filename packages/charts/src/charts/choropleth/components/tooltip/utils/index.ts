import calculatePartialPercents from './calculate-partials-percents';

export { calculatePartialPercents };
