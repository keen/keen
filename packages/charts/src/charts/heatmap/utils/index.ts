import { getTooltipLabel } from './get-tooltip-label';
import { generateBlocks } from './generate-blocks';

export { generateBlocks, getTooltipLabel };
