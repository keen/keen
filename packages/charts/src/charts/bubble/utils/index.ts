import { randomizeVector } from './animation.utils';
import { generateBubbles, bubbleColorScale } from './chart.utils';

export { randomizeVector, generateBubbles, bubbleColorScale };
