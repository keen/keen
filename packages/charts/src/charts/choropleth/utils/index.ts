import { fetchMapTopology, MapType } from './map.utils';
import { generateChoropleth, GeoPropety } from './chart.utils';

export { fetchMapTopology, generateChoropleth, MapType, GeoPropety };
