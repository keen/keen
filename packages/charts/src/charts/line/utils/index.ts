import {
  calculateStackData,
  calculateStackAreaData,
  sortKeys,
  calculateMaxMinSeriesValue,
} from './data.utils';
import {
  groupMarksByPosition,
  findMarksInCluster,
  showAllMarks,
  generateLineMarks,
} from './marks.utils';
import { generateSteps } from './steps.utils';
import { calculatePath } from './lines.utils';
import {
  generateSeriesBlockHeight,
  generateAreaGradient,
} from './gradient.utils';
import {
  generateGroupedLines,
  generateStackLines,
  generateLines,
} from './chart.utils';
import { calculateArea } from './area.utils';

export {
  calculateStackData,
  calculateStackAreaData,
  sortKeys,
  calculateMaxMinSeriesValue,
  groupMarksByPosition,
  findMarksInCluster,
  showAllMarks,
  generateLineMarks,
  generateSteps,
  calculatePath,
  generateSeriesBlockHeight,
  generateAreaGradient,
  generateGroupedLines,
  generateStackLines,
  generateLines,
  calculateArea,
};
