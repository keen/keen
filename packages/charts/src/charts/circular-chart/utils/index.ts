import { generateCircularChart } from './chart.utils';
import { getCircularChartTooltipContent } from './tooltip.utils';
import { createArcTween, animateArcPath } from './animate.utils';

export {
  generateCircularChart,
  getCircularChartTooltipContent,
  createArcTween,
  animateArcPath,
};
