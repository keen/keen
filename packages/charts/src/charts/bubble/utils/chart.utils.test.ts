import { generateBubbles } from './chart.utils';

import { chartFixture } from '../bubble-chart.fixtures';

describe('@keen.io/charts', () => {
  describe('<BubbleChart /> - utils', () => {
    describe('generateBubbles()', () => {
      it('should sort "bubbles" in descending order', () => {
        const { bubbles } = generateBubbles(chartFixture);

        expect(bubbles).toMatchSnapshot();
      });

      it('should calculate cartesian coordinates for center position', () => {
        const { middlePoint } = generateBubbles(chartFixture);

        expect(middlePoint).toMatchInlineSnapshot(`
          Array [
            44,
            50,
          ]
        `);
      });

      it('should create domain for "xScale"', () => {
        const { xScale } = generateBubbles(chartFixture);

        expect(xScale.domain()).toMatchInlineSnapshot(`
          Array [
            0,
            400,
          ]
        `);
      });

      it('should create domain for "yScale"', () => {
        const { yScale } = generateBubbles(chartFixture);

        expect(yScale.domain()).toMatchInlineSnapshot(`
          Array [
            0,
            500,
          ]
        `);
      });
    });
  });
});
