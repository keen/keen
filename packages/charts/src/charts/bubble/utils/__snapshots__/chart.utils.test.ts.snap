// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`@keen.io/charts <BubbleChart /> - utils generateBubbles() should sort "bubbles" in descending order 1`] = `
Array [
  Object {
    "color": "green",
    "key": "Google-2",
    "radius": 20,
    "selector": Array [
      2,
    ],
    "x": 31.25,
    "y": 76,
  },
  Object {
    "color": "blue",
    "key": "Facebook-0",
    "radius": 17.82857142857143,
    "selector": Array [
      0,
    ],
    "x": 88,
    "y": 0,
  },
  Object {
    "color": "yellow",
    "key": "Linkedin-1",
    "radius": 12.4,
    "selector": Array [
      1,
    ],
    "x": 20.5,
    "y": 80,
  },
  Object {
    "color": "orange",
    "key": "Youtube-3",
    "radius": 9.142857142857142,
    "selector": Array [
      3,
    ],
    "x": 80,
    "y": 36,
  },
]
`;
